modules/ocf_mirrors/files/collect-mirrors-stats

```diff
@@ -0,0 +1,168 @@
+            else:
+                projects['other']['up'] += int(stats[-2])
+                projects['other']['down'] += int(stats[-1])
+
+    return projects
+
+
+def process_rsync_log(filename):
+    """Parse Tx/Rx information out of an rsyncd log file.
+
+    This function is project-agnostic.
+    """
+
+    tx = 0
+    rx = 0
+    with filename.open('r') as f:
+        for line in f.readlines():
+            # log format is <datetime> <pid> sent <tx> bytes received <rx> bytes total size
+            if line.find('sent') > 0:
+                stats = line.split()
+                tx += int(stats[4])
+                rx += int(stats[7])
+
+    return tx, rx
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Process mirrors logs to calculate network usage '
+                                                 'and store in ocfstats')
+    parser.add_argument('-q', '--quiet', action='store_true',
+                        help='do not print stats after collecting them')
+    parser.add_argument('-n', '--dry-run', action='store_true',
+                        help='dry-run mode, do not write statistics to the database')
+    parser.add_argument('--apache', nargs='*',
+                        help='Apache log file(s) to process')
+    parser.add_argument('--rsync', nargs='*',
+                        help='rsync log files(s) to process')
+    parser.add_argument('date', nargs='?', default=date.today() - timedelta(1),
+                        help='date for use in filtering log entries')
+
+    args = parser.parse_args()
+
+    # first, figure out what projects we're getting stats for
+    sources = []
+    for project in MIRRORS_DATA_PATH.iterdir():
+        if project.is_dir() and not project.name.startswith('.'):
+            sources.append(project.name)
+
+    sources.append('other')  # catchall
+
+    # then, figure out which Apache logs we're pulling stats from
+    if args.apache:
+        apache_log_files = args.apache
+    else:
+        apache_log_files = []
+        for log_name in APACHE_LOG_FILES:
+            # logrotate rotates the log at 6am, but this script
+            # runs at midnight. So to capture 24h of data, we need
+            # to parse the rotated log as well. (the .1 file)
+            log = APACHE_LOG_PATH / log_name
+            apache_log_files.append(log)
+            apache_log_files.append(Path(str(log) + '.1'))
+
+    # and the same for rsync
+    if args.rsync:
+        rsync_log_files = args.rsync
+    else:
+        rsync_log_files = []
+        for log_file in RSYNC_LOG_PATH.iterdir():
+            # log files are named <project>.log
+            rsync_log_files.append(log_file)
+
+    # set up data structures
+    projects = {}
+    for project in sources:
+        projects[project] = {'up': 0, 'down': 0}
+
+    # pull the data itself from apache
+    for log_filename in apache_log_files:
+        projects = process_apache_log(projects, log_filename, args.date)
```

Comment:
Construct `projects` in `process_apache_log` (instead of above) and then just pass `sources` as a parameter to this function. That way we don't do this awkward thing where we pass in `projects` as a parameter, it's mutated, and then we assign it back to itself. (The assignment is currently a no-op.)

Reply:
I think that breaks the desired abstraction. we shouldn't construct `projects` from inside the function, because either we'll be: (a) overloading the function from 'just' processing a log file to having to know about how to construct the projects dictionary or (b) we won't know what a valid project is or isn't given just the log file itself. iteratively adding to the dictionary isn't ideal, but it's reminiscent of an iterative ETL approach and allows us to avoid having to do some kind of dictionary merging outside of the function.